[dynamo] lift the constraint that cannot make_fx a dynamo compiled function

ydwu4 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #109734
* #109916

Since we've added support for FakeTensor and SymBool inputs in dynamo, this PR removes the constraint that cannot make_fx a dynamo compiled function.

Checking the commit history, the is_fx_tracing flag in dynamo [code here](https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/eval_frame.py#L382C1-L383C1) seems to be mainly used to provide a nice error message for symbolic_trace. This PR changes how to detect we're during symbolic_trace by looking at is_fx_tracing flag and checking any arguments are proxies. This allows make_fx to be able to call into an optimized function.

**Test Plan:**
See added test.

 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

ydwu4 committed:
[dynamo] make symbool input propagate guard to outter shape_env while dynamo tracing

[ghstack-poisoned]

ydwu4 committed:
Update on "[dynamo] make symbool input propagate guard to outter shape_env while dynamo tracing"


Before this PR, for symbool inputs, dynamo creates a new symbool in its own shape_env and didn't propagate the guards to the input's shape_env. This causes diverged behavior between make_fx over a dynamo compiled function  vs make_fx over an eager run function. For the following code:
```python
        def fn(pred, x):
            if pred:
                return x.sin()
            else:
                return x.cos()

        def symbool_fn_compiled(x):
            sym_bool = x.shape[0] == 3
            return torch.compile(backend="eager", fullgraph=True)(fn)(sym_bool, x)

        def symbool_fn_not_compiled(x):
            pred = x.shape[0] == 3
            return fn(pred, x)

        x = torch.ones(3, 4)
        x2 = torch.ones(4, 3)

        compiled_gm = make_fx(
            lambda x: symbool_fn_compiled(x), tracing_mode="symbolic"
        )(x)
        compiled_gm_str = compiled_gm.print_readable(print_output=False)

        not_compiled_gm = make_fx(
            lambda x: symbool_fn_not_compiled(x), tracing_mode="symbolic"
        )(x)
        not_compiled_gm_str = not_compiled_gm.print_readable(print_output=False)
```
, we'll see compiled_gm.shape_env.guards is empty but not_compiled_gm.shape_env.guards has a guard [Eq(s0, 3)]. This also causes the gm_str of these two looks a little different:
```python
# not_compiled_gm_str
# s0 is specialized as 3
class <lambda>(torch.nn.Module):
    def forward(self, x_1: f32[3, s1]):
        # No stacktrace found for following nodes
        sin: f32[3, s1] = torch.ops.aten.sin.default(x_1);  x_1 = None
        return sin

# compiled_gm_str
# s0 is not specialized
class <lambda>(torch.nn.Module):
    def forward(self, x_1: f32[s0, s1]):
        # No stacktrace found for following nodes
        sin: f32[s0, s1] = torch.ops.aten.sin.default(x_1);  x_1 = None
        return sin
```

This PR fixes the inconsistency by propagating the guard to outter shape_env when dynamo is tracing. This is achieved by introducing a InputSymNodeVariable that tracks (origin_sym_num, dynamo_sym_num) and whenever eva_expr is called, both symbols will be evaluated and properly guarded.

Test Plan:
Modified existing tests to reflect the change.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ydwu4 commented:
The current implementation didn't handle expressions that are generated from input symbool correctly: to do that, we'll need to propagate the expressions on both outer shape_env and inner shape_env. It's requires a lot of effort.

The next plan is run the generated shape_env guard once with input symbol (from outer shape_env) to install the guards in outer shape_env

ezyang commented:
do you want review now or later

ydwu4 commented:
> do you want review now or later

Later today! I got some ideas. Will ping you when it's ready

ydwu4 committed:
Update on "[WIP][dynamo] make symbool input propagate guard to outer shape_env while dynamo tracing"


Before this PR, for symbool inputs, dynamo creates a new symbool in its own shape_env and didn't propagate the guards to the input's shape_env. This causes diverged behavior between make_fx over a dynamo compiled function  vs make_fx over an eager run function. For the following code:
```python
        def fn(pred, x):
            if pred:
                return x.sin()
            else:
                return x.cos()

        def symbool_fn_compiled(x):
            sym_bool = x.shape[0] == 3
            return torch.compile(backend="eager", fullgraph=True)(fn)(sym_bool, x)

        def symbool_fn_not_compiled(x):
            pred = x.shape[0] == 3
            return fn(pred, x)

        x = torch.ones(3, 4)
        x2 = torch.ones(4, 3)

        compiled_gm = make_fx(
            lambda x: symbool_fn_compiled(x), tracing_mode="symbolic"
        )(x)
        compiled_gm_str = compiled_gm.print_readable(print_output=False)

        not_compiled_gm = make_fx(
            lambda x: symbool_fn_not_compiled(x), tracing_mode="symbolic"
        )(x)
        not_compiled_gm_str = not_compiled_gm.print_readable(print_output=False)
```
, we'll see compiled_gm.shape_env.guards is empty but not_compiled_gm.shape_env.guards has a guard [Eq(s0, 3)]. This also causes the gm_str of these two looks a little different:
```python
# not_compiled_gm_str
# s0 is specialized as 3
class <lambda>(torch.nn.Module):
    def forward(self, x_1: f32[3, s1]):
        # No stacktrace found for following nodes
        sin: f32[3, s1] = torch.ops.aten.sin.default(x_1);  x_1 = None
        return sin

# compiled_gm_str
# s0 is not specialized
class <lambda>(torch.nn.Module):
    def forward(self, x_1: f32[s0, s1]):
        # No stacktrace found for following nodes
        sin: f32[s0, s1] = torch.ops.aten.sin.default(x_1);  x_1 = None
        return sin
```

This PR fixes the inconsistency by propagating the guard to outter shape_env when dynamo is tracing. This is achieved by introducing a InputSymNodeVariable that tracks (origin_sym_num, dynamo_sym_num) and whenever eva_expr is called, both symbols will be evaluated and properly guarded.

Test Plan:
Modified existing tests to reflect the change.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ydwu4 committed:
Update on "[dynamo] lift the constraint that cannot make_fx a dynamo compiled function"


Since we've added support for FakeTensor and SymBool inputs in dynamo, this PR removes the constraint that cannot make_fx a dynamo compiled function.

Checking the commit history, the is_fx_tracing flag in dynamo [code here](https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/eval_frame.py#L382C1-L383C1) seems to be mainly used to provide a nice error message for symbolic_trace. This PR changes how to detect we're during symbolic_trace by looking at is_fx_tracing flag and checking any arguments are proxies. This allows make_fx to be able to call into an optimized function.

**Test Plan:**
See added test.

 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ydwu4 committed:
Update on "[dynamo] lift the constraint that cannot make_fx a dynamo compiled function"


Since we've added support for FakeTensor and SymBool inputs in dynamo, this PR removes the constraint that cannot make_fx a dynamo compiled function.

Checking the commit history, the is_fx_tracing flag in dynamo [code here](https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/eval_frame.py#L382C1-L383C1) seems to be mainly used to provide a nice error message for symbolic_trace. This PR changes how to detect we're during symbolic_trace by looking at is_fx_tracing flag and checking any arguments are proxies. This allows make_fx to be able to call into an optimized function.

**Test Plan:**
See added test.

 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ydwu4 committed:
Update on "[dynamo] lift the constraint that cannot make_fx a dynamo compiled function"


Since we've added support for FakeTensor and SymBool inputs in dynamo, this PR removes the constraint that cannot make_fx a dynamo compiled function.

Checking the commit history, the is_fx_tracing flag in dynamo [code here](https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/eval_frame.py#L382C1-L383C1) seems to be mainly used to provide a nice error message for symbolic_trace. This PR changes how to detect we're during symbolic_trace by looking at is_fx_tracing flag and checking any arguments are proxies. This allows make_fx to be able to call into an optimized function.

**Test Plan:**
See added test.

 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ydwu4 committed:
Update on "[dynamo] lift the constraint that cannot make_fx a dynamo compiled function"


Since we've added support for FakeTensor and SymBool inputs in dynamo, this PR removes the constraint that cannot make_fx a dynamo compiled function.

Checking the commit history, the is_fx_tracing flag in dynamo [code here](https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/eval_frame.py#L382C1-L383C1) seems to be mainly used to provide a nice error message for symbolic_trace. This PR changes how to detect we're during symbolic_trace by looking at is_fx_tracing flag and checking any arguments are proxies. This allows make_fx to be able to call into an optimized function.

**Test Plan:**
See added test.

 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ydwu4 committed:
Update on "[dynamo] lift the constraint that cannot make_fx a dynamo compiled function"


Since we've added support for FakeTensor and SymBool inputs in dynamo, this PR removes the constraint that cannot make_fx a dynamo compiled function.

Checking the commit history, the is_fx_tracing flag in dynamo [code here](https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/eval_frame.py#L382C1-L383C1) seems to be mainly used to provide a nice error message for symbolic_trace. This PR changes how to detect we're during symbolic_trace by looking at is_fx_tracing flag and checking any arguments are proxies. This allows make_fx to be able to call into an optimized function.

**Test Plan:**
See added test.

 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ezyang commented on torch/_dynamo/eval_frame.py:
```diff
@@ -377,9 +378,14 @@ def get_compiler_config():
             if (
                 not isinstance(self, DisableContext)
                 and torch.fx._symbolic_trace.is_fx_tracing()
+                and any_arg_is_proxy
```
What's going on here?

ydwu4 replied:
is_fx_tracing flag is also set for make_fx. So we try to detect whether we're in symbolic_trace mode by checking if any arguments are wrapped as proxy.

ezyang commented:
I am not so hot on just removing this assert. Maybe we can treat "eager" dynamo compiled regions differently from conventional (compiled) regions?

ydwu4 commented:
> I am not so hot on just removing this assert. Maybe we can treat "eager" dynamo compiled regions differently from conventional (compiled) regions?

Yeah, sounds good. Let me fix it by adding a backend check as well.

ydwu4 committed:
Update on "[dynamo] lift the constraint that cannot make_fx a dynamo compiled function"


Since we've added support for FakeTensor and SymBool inputs in dynamo, this PR removes the constraint that cannot make_fx a dynamo compiled function.

Checking the commit history, the is_fx_tracing flag in dynamo [code here](https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/eval_frame.py#L382C1-L383C1) seems to be mainly used to provide a nice error message for symbolic_trace. This PR changes how to detect we're during symbolic_trace by looking at is_fx_tracing flag and checking any arguments are proxies. This allows make_fx to be able to call into an optimized function.

**Test Plan:**
See added test.

 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ydwu4 committed:
Update on "[dynamo] lift the constraint that cannot make_fx a dynamo compiled function"


Since we've added support for FakeTensor and SymBool inputs in dynamo, this PR removes the constraint that cannot make_fx a dynamo compiled function.

Checking the commit history, the is_fx_tracing flag in dynamo [code here](https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/eval_frame.py#L382C1-L383C1) seems to be mainly used to provide a nice error message for symbolic_trace. This PR changes how to detect we're during symbolic_trace by looking at is_fx_tracing flag and checking any arguments are proxies. This allows make_fx to be able to call into an optimized function.

**Test Plan:**
See added test.

 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ezyang commented on test/dynamo/test_misc.py:
```diff
@@ -4462,6 +4465,66 @@ def f(x):
+            for gm in gms[1:]:
+                self.assertExpectedInline(
+                    _get_guard_exprs(gm),
+                    """Eq(Piecewise((1, Eq(s0, 3)), (0, True)), 1)""",
```
The RHS of the Piecewise is constant, why isn't this simplified 🤔

ezyang reviewed (changes requested):
Please more carefully review your expect test output, it still looks wrong

ydwu4 committed:
Update on "[dynamo] lift the constraint that cannot make_fx a dynamo compiled function"


Since we've added support for FakeTensor and SymBool inputs in dynamo, this PR removes the constraint that cannot make_fx a dynamo compiled function.

Checking the commit history, the is_fx_tracing flag in dynamo [code here](https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/eval_frame.py#L382C1-L383C1) seems to be mainly used to provide a nice error message for symbolic_trace. This PR changes how to detect we're during symbolic_trace by looking at is_fx_tracing flag and checking any arguments are proxies. This allows make_fx to be able to call into an optimized function.

**Test Plan:**
See added test.

 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ydwu4 committed:
Update on "[dynamo] lift the constraint that cannot make_fx a dynamo compiled function"


Since we've added support for FakeTensor and SymBool inputs in dynamo, this PR removes the constraint that cannot make_fx a dynamo compiled function.

Checking the commit history, the is_fx_tracing flag in dynamo [code here](https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/eval_frame.py#L382C1-L383C1) seems to be mainly used to provide a nice error message for symbolic_trace. This PR changes how to detect we're during symbolic_trace by looking at is_fx_tracing flag and checking any arguments are proxies. This allows make_fx to be able to call into an optimized function.

**Test Plan:**
See added test.

 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ydwu4 committed:
Update on "[dynamo] lift the constraint that cannot make_fx a dynamo compiled function"


Since we've added support for FakeTensor and SymBool inputs in dynamo, this PR removes the constraint that cannot make_fx a dynamo compiled function.

Checking the commit history, the is_fx_tracing flag in dynamo [code here](https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/eval_frame.py#L382C1-L383C1) seems to be mainly used to provide a nice error message for symbolic_trace. This PR changes how to detect we're during symbolic_trace by looking at is_fx_tracing flag and checking any arguments are proxies. This allows make_fx to be able to call into an optimized function.

**Test Plan:**
See added test.

 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ydwu4 committed:
Update on "[dynamo] lift the constraint that cannot make_fx a dynamo compiled function"


Since we've added support for FakeTensor and SymBool inputs in dynamo, this PR removes the constraint that cannot make_fx a dynamo compiled function.

Checking the commit history, the is_fx_tracing flag in dynamo [code here](https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/eval_frame.py#L382C1-L383C1) seems to be mainly used to provide a nice error message for symbolic_trace. This PR changes how to detect we're during symbolic_trace by looking at is_fx_tracing flag and checking any arguments are proxies. This allows make_fx to be able to call into an optimized function.

**Test Plan:**
See added test.

 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ydwu4 committed:
Update on "[dynamo] lift the constraint that cannot make_fx a dynamo compiled function"


Since we've added support for FakeTensor and SymBool inputs in dynamo, this PR removes the constraint that cannot make_fx a dynamo compiled function.

Checking the commit history, the is_fx_tracing flag in dynamo [code here](https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/eval_frame.py#L382C1-L383C1) seems to be mainly used to provide a nice error message for symbolic_trace. This PR changes how to detect we're during symbolic_trace by looking at is_fx_tracing flag and checking any arguments are proxies. This allows make_fx to be able to call into an optimized function.

**Test Plan:**
See added test.

 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ydwu4 committed:
Update on "[dynamo] lift the constraint that cannot make_fx a dynamo compiled function"


Since we've added support for FakeTensor and SymBool inputs in dynamo, this PR removes the constraint that cannot make_fx a dynamo compiled function.

Checking the commit history, the is_fx_tracing flag in dynamo [code here](https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/eval_frame.py#L382C1-L383C1) seems to be mainly used to provide a nice error message for symbolic_trace. This PR changes how to detect we're during symbolic_trace by looking at is_fx_tracing flag and checking any arguments are proxies. This allows make_fx to be able to call into an optimized function.

**Test Plan:**
See added test.

 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ydwu4 committed:
Update on "[dynamo] lift the constraint that cannot make_fx a dynamo compiled function"


Since we've added support for FakeTensor and SymBool inputs in dynamo, this PR removes the constraint that cannot make_fx a dynamo compiled function.

Checking the commit history, the is_fx_tracing flag in dynamo [code here](https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/eval_frame.py#L382C1-L383C1) seems to be mainly used to provide a nice error message for symbolic_trace. This PR changes how to detect we're during symbolic_trace by looking at is_fx_tracing flag and checking any arguments are proxies. This allows make_fx to be able to call into an optimized function.

**Test Plan:**
See added test.

 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ydwu4 committed:
Update on "[dynamo] lift the constraint that cannot make_fx a dynamo compiled function"


Since we've added support for FakeTensor and SymBool inputs in dynamo, this PR removes the constraint that cannot make_fx a dynamo compiled function.

Checking the commit history, the is_fx_tracing flag in dynamo [code here](https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/eval_frame.py#L382C1-L383C1) seems to be mainly used to provide a nice error message for symbolic_trace. This PR changes how to detect we're during symbolic_trace by looking at is_fx_tracing flag and checking any arguments are proxies. This allows make_fx to be able to call into an optimized function.

**Test Plan:**
See added test.

 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ydwu4 committed:
Update on "[dynamo] lift the constraint that cannot make_fx a dynamo compiled function"


Since we've added support for FakeTensor and SymBool inputs in dynamo, this PR removes the constraint that cannot make_fx a dynamo compiled function.

Checking the commit history, the is_fx_tracing flag in dynamo [code here](https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/eval_frame.py#L382C1-L383C1) seems to be mainly used to provide a nice error message for symbolic_trace. This PR changes how to detect we're during symbolic_trace by looking at is_fx_tracing flag and checking any arguments are proxies. This allows make_fx to be able to call into an optimized function.

**Test Plan:**
See added test.

 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ydwu4 committed:
Update on "[dynamo] lift the constraint that cannot make_fx a dynamo compiled function"


Since we've added support for FakeTensor and SymBool inputs in dynamo, this PR removes the constraint that cannot make_fx a dynamo compiled function.

Checking the commit history, the is_fx_tracing flag in dynamo [code here](https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/eval_frame.py#L382C1-L383C1) seems to be mainly used to provide a nice error message for symbolic_trace. This PR changes how to detect we're during symbolic_trace by looking at is_fx_tracing flag and checking any arguments are proxies. This allows make_fx to be able to call into an optimized function.

**Test Plan:**
See added test.

 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]
